[Liminal UI] Implement overlay, falloff and smaller design changes

yangchristina opened the pull request:
Resolves #3230



https://github.com/user-attachments/assets/1938af77-f70a-46a2-a229-0aa7d7ecc8dc

yangchristina commented:
@fbmcipher Thanks! Yes, I'll try both and send them to you once done.

@raineorshine Is using MUI a requirement? The MUI Swipeable Drawer doesn’t seem to support the behaviour we’re aiming for, and even if I get it working, the solution would likely be quite hacky.

yangchristina commented:
@fbmcipher here is fade only:


https://github.com/user-attachments/assets/1049bd81-91ac-4880-8dcc-2112f3c53685

yangchristina commented:
@fbmcipher Here is Fade & slide. I made the fade and slide transitions match the slide transition of the MUI drawer, but we can alway change the easing for just the fade transition.

https://github.com/user-attachments/assets/878176a6-ac26-4a99-8717-2e069d3833a5

raineorshine commented:
It's not ideal to reference a ref value in a React render function, because it's nonreactive. It won't re-render if the height changes, and the particular value that it references on a given render is going to implicitly depend on the timing of the component lifecycle and state updates. The value will end up either being outdated sometimes, or fortuitously never outdated, which is not something you want to rely on.

I don't have an immediate alternative in mind, but maybe you and @fbmcipher can come up with a more conventional solution.

yangchristina commented:
> It's not ideal to reference a ref value in a React render function, because it's nonreactive. It won't re-render if the height changes, and the particular value that it references on a given render is going to implicitly depend on the timing of the component lifecycle and state updates. The value will end up either being outdated sometimes, or fortuitously never outdated, which is not something you want to rely on.
> 
> I don't have an immediate alternative in mind, but maybe you and @fbmcipher can come up with a more conventional solution.

@raineorshine I can use useState to store the height instead and a ResizeObserver, that way it will update when the height changes. I know this is still not ideal, but the progressive blur effect doesn't work when it is stored inside SwipeableDrawer.
It doesn't blur the below thoughts when it is within SwipeableDrawer, it only worked when I moved it outside. This blur must match the height of the CommandMenu, hence why I need to track the height of CommandMenu.

fbmcipher commented:
@yangchristina Fade and slide works! Let's keep things as simple as possible for this iteration. If the animation still feels off, we can take another look later in a separate piece of work.

I'll field some of the more technical questions tomorrow.

yangchristina commented:
> @yangchristina Fade and slide works! Let's keep things as simple as possible for this iteration. If the animation still feels off, we can take another look later in a separate piece of work.
> 
> I'll field some of the more technical questions tomorrow.

Thank you! I've implemented the fade and slide, and updated the video in the PR description.

fbmcipher reviewed (changes requested):
@yangchristina This looks great, and fade & slide looks great too! The light gradually fading in as the drawer opens feels much better than the janky animation we had before.

During testing, I did notice one detail that feels missing: the backdrop doesn't fade out _as_ you close the drawer with your finger.

https://github.com/user-attachments/assets/12261fc1-7649-4eb3-9033-ee7a8f834833

However, I accept that this probably isn't easy to do in a performant and elegant way with MUI's `SwipeableDrawer` limitations.

Between `fade` and `fade+slide`, I actually preferred `fade` – but I couldn't advocate for its implementation given the complexity and hackiness it'd take to work against MUI.

I like the idea of removing our dependency on MUI's `SwipeableDrawer` and finding a more customisable component/hook so we can customise further, but I think that work would be better done in another PR.

For now, this looks and works great IMO!

> Actually, I just found out that the culprit for the progressive blur not working isn't the portal, but actually mix-blend-mode. It was working when outside the drawer because it wasn't affected by mix-blend-mode: screen on the overlay there. I will look more into this.

The likely culprit here: when we use `mix-blend-mode`, a new [stacking context](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_positioned_layout/Stacking_context) is created. That changes the element’s compositing order and isolation, which can break or change the expected blending with siblings and ancestors. This behaviour is worth keeping in mind as our designs use a lot of blend modes.

Just a few minor changes from me!

yangchristina commented on src/durations.config.ts:
```diff
@@ -37,6 +37,8 @@ const durationsConfig = {
   /* Duration for context view disappearing text animations */
   disappearingUpperRight: 500,
   disappearingLowerLeft: 500,
+  /* Duration for command menu swipable drawer animation. */
```
The reason I added this is to match MUI's easing. Because of the implementation of `FadeTransition`, it requires the variant to also exist in the durationConfig object. To remove it from durationConfig means I'd have to do something like:
```
<CSSTransition
      key={id}
      classNames={fadeClasses}
      timeout={durations.get(type === 'commandMenuDrawer' ? 'medium' : type)}
```
This doesn't seem ideal, so adding to the durationConfig seems like a better solution to me.
Please let me know if you have a better solution!

fbmcipher replied:
Thank you for explaining – that makes total sense now!

All things considered, it seems like you've arrived at the most ideal solution. Happy with this :)

yangchristina commented on src/components/CommandMenu/CommandMenu.tsx:
```diff
@@ -85,93 +110,147 @@ const CommandMenu = () => {
+                      fontSize: fontSize * 0.85,
+                    }}
+                  >
+                    <CloseIconV2 /> Done
```
Updated! See video below for how it looks now.

yangchristina commented on src/components/icons/CloseIconV2.tsx:
```diff
@@ -0,0 +1,28 @@
+import IconType from '../../@types/IconType'
+
+/** Close icon. */
+const CloseIconV2 = ({ fill, width = 10, height = 11, cssRaw, style }: IconType) => {
```
removed thanks! [2111e93](https://github.com/cybersemics/em/pull/3246/commits/2111e9363be6ef80908054335ea73481660de2f3)

fbmcipher commented on src/components/CommandMenu/CommandMenu.tsx:
```diff
@@ -85,93 +110,147 @@ const CommandMenu = () => {
+                      zIndex: 'modal',
+                    })}
+                    style={{
+                      fontSize: fontSize * 0.85,
```
Thank you! 

> The only potential concern is that changing a font-size on a parent could cause unexpected side-effects

Good point! Let's keep an eye out for that. Looking through the codebase, the majority of fontSize styles use `em`, so let's standardise on `em` for now.

If we start seeing unwanted cascading side effects, we can revisit and move to `rem` at that point.

yangchristina replied:
Sounds good! I think there was talk to move to `rem` at some point, but I didn't see it ever happened.

fbmcipher reviewed (approved):
Thanks for the changes and explanations!

> Thanks! Turns out MUI's exit transition easing is different from the enter transition. I've updated the easing here:
https://github.com/cybersemics/em/commit/f545e661606a65e91f7d5d25ddf4a4c756030020. How is this?

The new easing on close feels great – glad we're fully consistent with MUI wherever possible.

What I actually meant, though (I could have visualised better in the video) – is that the glow/overlay doesn't fade out as the Command Center is closed **gesturally**:


https://github.com/user-attachments/assets/91f893cd-5b4e-4eae-ac09-f9be9243ab27



It looks great when it closes via the Done button or other tap interaction – the glow fades  out nicely as the panel slides down.

But if the user slides down with their finger, then the overlay _doesn't_ fade out until we reach the bottom. The idea of light sliding like this doesn't *feel* quite right – light doesn't move like that IRL.

Though I fully appreciate that any solution would likely be inelegant and potentially introduce perf issues due the limitations we discussed around MUI, so I'm happy to circle back around to this later/in a separate issue.

---

Everything else looks great to me! I'm happy to approve and pass on to @raineorshine for one last check before we merge.

As soon as it's merged, we can move onto the next issue!

I'm really happy with how true to the design this PR is!

yangchristina commented:
> What I actually meant, though (I could have visualised better in the video) – is that the glow/overlay doesn't fade out as the Command Center is closed **gesturally**:

I see, thanks for explaining! Yes, the fade amount currently isn't tied to the height of the drawer, and MUI doesn't change the value of `showCommandMenu` until you let go, so the fade won't begin until you let go. Unfortunately I don't see a way to track the transform value of the drawer right now. I would be happy to circle back to this on a later issue as well.

yangchristina commented:
> Thanks for thinking ahead. Let's do it.
> 
> 1. Just fix the uncanny gap, then merge this PR.
> 2. Migrate to [Vaul](https://vaul.emilkowal.ski/) or [react-spring-bottom-sheet](https://react-spring.bottom-sheet.dev/) as a refactor in a new PR. Ignore background glow behavior, but otherwise ensure no behavioral changes from current implementation.
> 3. Implement the fade animation and background glow behavior as desired.

@raineorshine Sounds good! I've removed the margin-top, as suggested:


3d12c82

<img width="422" height="766" alt="Screenshot 2025-10-16 at 11 24 11 PM" src="https://github.com/user-attachments/assets/e6505af0-baa3-4872-9cd8-3e249434f83d" />

yangchristina commented:
@raineorshine would it be fine to implement the buttons before migrating the drawer library? I already have the changes ready, so would be easier to get those in first.

fbmcipher commented:
@yangchristina @raineorshine Just a heads-up: I opened #3283 with more details on the new drawer implementation, based on the upcoming stages of Command Center development.

yangchristina commented:
@raineorshine is there anything else needed for this PR?

Since this hasn't been merged yet, and I need this branch as the base for my next PR, I've created the PR on my forked repository for now, since I didn't see a way to do it on the main repository. 

https://github.com/yangchristina/em/pull/270#issue-3534399705

@fbmcipher how is this for active styling? (Sorry I couldn't mention you from my forked repository, so am asking here)

raineorshine commented:
> @raineorshine is there anything else needed for this PR?

Just waiting for @JoeDuong-dev to test!

> Since this hasn't been merged yet, and I need this branch as the base for my next PR, I've created the PR on my forked repository for now, since I didn't see a way to do it on the main repository. 

You can branch off your other branch and open a PR here. The diff will show all the changes, but I can select from the commits dropdown to limit the diff to just the most recent batch of work. Then when this PR gets merged, we can update the base.

yangchristina commented on src/components/CommandMenu/CommandMenu.tsx:
```diff
@@ -85,93 +105,144 @@ const CommandMenu = () => {
-                <MultiselectMessage />
-                <button
-                  {...fastClick(onClose)}
+      <>
```
Actually let me undo that commit, something seems to have gone wrong there.

yangchristina commented:
@raineorshine I think this is a flaky test?
<img width="2400" height="600" alt="drag-should-show-alert-and-quick-drop-panels-when-long-pressing-after-drag-operation-1-diff" src="https://github.com/user-attachments/assets/44c19fa4-1c74-48a6-a192-3b0e92007671" />

raineorshine commented:
That seems like a reasonable theory since this PR doesn't touch thought layout. We eliminated all flakiness in the tests earlier this year, so if this is indeed flaky then it's a recent regression.

@karunkop is in charge of the test suite, so he can investigate further and provide guidance on how to move forward.

yangchristina commented:
@karunkop The tests appear to be passing for me locally. The only change I made was to remove a react fragment, therefore it shouldn't have caused any tests to break. I just merged with main to see if that was the issue, but it looks like the test failure is consistent.

I tried running puppeteer locally, and got passing tests.

<img width="315" height="115" alt="Screenshot 2025-10-26 at 6 07 41 PM" src="https://github.com/user-attachments/assets/647b4bcb-0c8c-42ba-b4a8-4698a068f887" />

Do you have an idea of what the issue could be?

yangchristina commented:
@raineorshine tests are passing now after merging with main.
